File: src/algorithm/leetcode/Problem160.java
package algorithm.leetcode;

public class Problem160 {

    public static ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if (headA == null || headB == null)
            return null;
        int lenA = 0, lenB = 0;
        ListNode pa = headA, pb = headB;
        while (pa != null) {
            pa = pa.next;
            lenA++;
        }
        while (pb != null) {
            pb = pb.next;
            lenB++;
        }
        pa = headA;
        pb = headB;
        if (lenA > lenB) {
            for (int i = 0; i < lenA - lenB; i++) {
                pa = pa.next;
            }
        } else {
            for (int i = 0; i < lenB - lenA; i++) {
                pb = pb.next;
            }
        }
        while (pa != null && pb != null) {
            if (pa == pb)
                return pa;
            pa = pa.next;
            pb = pb.next;
        }
        return null;
    }

}

File: src/algorithm_cpp/problem160.cpp
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
	if (headA == nullptr || headB == nullptr) return nullptr;
        ListNode *end = nullptr;
	ListNode *ptr1 = headA, *ptr2 = headB;
	while (ptr1 != ptr2) {
	    if (ptr1->next != nullptr) {
		ptr1 = ptr1->next;
	    } else {
		if (end != nullptr && ptr1 != end) return nullptr;
		end = ptr1;
		ptr1 = headB;
	    }
	    if (ptr2->next != nullptr) {
		ptr2 = ptr2->next;
	    } else {
		if (end != nullptr && ptr2 != end) return nullptr;
		end = ptr2;
		ptr2 = headA;
	    }
	}
	return ptr1;
    }
};
